feature(admin-server,auth-server): Refactor to support querying account subscriptions from admin-server

dschom opened the pull request:
## Because

- We need to be able to display account subscription information in the admin panel.

## This pull request

- Establishes a pattern of extracting base classes in order to facility shared routines across work-spaces, which in this case is the admin-server and auth-server.
- Move routines that provide access to stripe subscription information to fxa-shared/payments
- Move routines and types that provide access to iap subscriptions to fxa-shared/payments
- Move common routines for reporting to errors to sentry to fxa-shared/sentry

## Issue that this pull request solves

Closes: # (issue number)

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [ ] I have added necessary documentation (if appropriate).
- [ ] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

Please attach the screenshots of the changes made in case of change in user interface.

## Other information (Optional)

Any other information that is important to this pull request.

dschom commented on packages/fxa-auth-server/lib/payments/configuration/manager.ts:
```diff
@@ -1,143 +1,40 @@
+export class PaymentConfigManager extends PaymentConfigManagerBase {
   constructor() {
     const config = Container.get(AppConfig);
-    this.prefix = `${config.authFirestore.prefix}payment-config-`;
```
Deleted code typically has been hoisted to base class in fxa-shared.

dschom commented on packages/fxa-auth-server/lib/payments/configuration/manager.ts:
```diff
@@ -1,143 +1,40 @@
 import { AppConfig, AuthFirestore, AuthLogger } from '../../types';
+
+import {
+  PaymentConfigManager as PaymentConfigManagerBase,
```
I kept the class names in fxa-shared the same and then do a 'as' here. Not sure if this good style or not. Let me know your thoughts. Happy to change this if needed.

dschom commented on packages/fxa-shared/sentry/index.ts:
```diff
@@ -1,3 +1,68 @@
+ * @param {string | import('@hapi/joi').ValidationError} error
+ */
+export function reportValidationError(
+  sentry: Pick<Hub, 'withScope' | 'captureMessage'>,
```
Type might seem a bit strange here, but see comment in doc about being context agnostic. Essentially, we don't want to make assumptions about the type of sentry instance (e.g. that is a '@sentry/node' instance).

bbangert replied:
We're only using Sentry as a module global, are we expecting to pass around confined instances of it now?

dschom replied:
See other comment about making sentry context agnostic. Maybe we can side bar on this? I agree this one off approach is not ideal.

dschom replied:
I was able to work around the context issue by moving these functions out of index.ts.

dschom commented on packages/fxa-auth-server/lib/payments/iap/apple-app-store/subscription-purchase.ts:
```diff
@@ -1,208 +1,5 @@
-    return this.offerType === OfferType.Introductory;
-  }
-}
+export * from 'fxa-shared/payments/iap/apple-app-store/subscription-purchase';
```
I took this approach in an effort to minimize side effects of the refactor and to cut down on changes to existing import statements. I can also see decorating this module with auth-server only functionality down the road.

bbangert replied:
Ah, ok, so later we could move these entirely to fxa-shared as well as the tests.

dschom commented on packages/fxa-auth-server/lib/sentry.js:
```diff
@@ -276,51 +278,6 @@ function ignoreErrors(error) {
   return IGNORED_ERROR_NUMBERS.includes(error.errno);
 }
 
-/**
```
Moved to fxa-shared/sentry

bbangert commented on packages/fxa-auth-server/lib/payments/iap/google-play/user-manager.ts:
```diff
@@ -21,27 +21,23 @@ import Container from 'typedi';
   constructor(
-    private purchasesDbRef: CollectionReference,
-    private purchaseManager: PurchaseManager
+    public purchasesDbRef: CollectionReference,
```
Why is this one public?

dschom replied:
Ah good catch, let me try making that private again.

bbangert commented on packages/fxa-auth-server/lib/payments/stripe.ts:
```diff
@@ -181,40 +147,33 @@ class NoopCacheClearStrategy implements CacheClearStrategy {
-    this.webhookSecret = config.subscriptions.stripeWebhookSecret;
-    this.taxIds = config.subscriptions.taxIds;
-    this.currencyHelper = Container.get(CurrencyHelper);
+    super(config, statsd, log);
```
Dependency on config format again, ideally these should have their own config formats.

bbangert replied:
I see now that this is like the above bit, and the Stripe object has its own type that config happens to meet.

bbangert commented on packages/fxa-auth-server/lib/payments/stripe.ts:
```diff
@@ -1582,66 +1528,11 @@ export class StripeHelper {
-        `Stripe Customer: ${customer.id} has mismatched uid in metadata.`
-      );
+    try {
+      return await super.fetchCustomer(uid, expand);
```
This isn't `this.fetchCustomer`? There's two implementations?

dschom replied:
It's hijacking the base's function. If I were to call this.fetchCustomer it would just be a never ending recursive call and would blow up the call stack.

bbangert commented on packages/fxa-auth-server/lib/routes/subscriptions/stripe-webhook.ts:
```diff
@@ -646,7 +646,7 @@ export class StripeWebhookHandler extends StripeHandler {
         plan,
       });
-      reportValidationError(msg, error as any);
+      reportValidationError(Sentry, msg, error as any);
```
Sentry is a global, is there any point in passing it around as it'll be the same everywhere when imported?

dschom replied:
I left a comment in fxa-shared/sentry/index.ts Sentry about this. Basically this allows the Sentry instance to be context agnostic. 

This change was a direct result of running into two issues. First, if I import '@sentry/node' in fxa-shared it ended up causing problems on the client side. This is because fxa-shared/sentry is referenced from a client side context in some places. When @sentry/node is require from a browser context it results in a runtime error. Second, if ignored the first issue and just referenced the global @sentry/node instance in fxa-shared, for some reason the mocks on Sentry in our tests all started failing. 

If this is a no go let me know, and I'd be happy to revisit it, as I am not completely happy with the solution. The browser context issue is easy enough to work around, but the way the mock stopped working was definitely puzzling and might turn into a time suck.

bbangert replied:
Ugh, ok, yea, that's annoying. Thanks for the context on that.

bbangert commented on packages/fxa-shared/payments/configuration/manager.ts:
```diff
@@ -0,0 +1,189 @@
+ * shared across workspaces.
+ */
+export abstract class PaymentConfigManager {
+  public productConfigDbRef: TypedCollectionReference<ProductConfig>;
+  public planConfigDbRef: TypedCollectionReference<PlanConfig>;
```
Were these public before?

dschom replied:
I'll double check...

dschom commented on packages/fxa-auth-server/lib/payments/configuration/manager.ts:
```diff
@@ -1,143 +1,40 @@
-    );
+    const firestore = Container.get(AuthFirestore);
+    const log = Container.get(AuthLogger) as any as Logger;
+    super(config, firestore, log);
```
Are you saying just copy what's needed into a new object? I am not sure I see why passing the config object is an issue. 

By the way, although there is shared object here, there is not a type level dependency. The base class's constructor defines it's own interface for a config object thereby decoupling it from the auth-server's config type.

dschom commented on packages/fxa-auth-server/lib/payments/iap/google-play/user-manager.ts:
```diff
@@ -57,19 +53,7 @@ export class UserManager {
-      if (packageName) {
-        query = query.where('packageName', '==', packageName);
-      }
+      let query = super.buildSubscriptionQuery(userId, sku, packageName);
```
I think this a misunderstanding on my part. I was thinking, partially from our discussion, that in the admin panel the raw result of the query would suffice. It seems like you are saying that this logic can't really be separated though, which as I look at this again makes sense. If this is case, I can hoist the entire method.

By the way, the other motivation for splitting this was that it meant the base class didn't need a reference to the purchase manager. Not a really big deal, but it's nice to drop a dependency when possible.

bbangert replied:
Ah, yes, these methods in the purchase-manager resolve quite a bit of subscription state, so they need to run.

dschom replied:
Sounds good. Not a biggie, but looking at this, it will definitely result in a decent chunk of code getting hoisted over to fxa-shared.

dschom replied:
Alright should be good to go now.

bbangert commented on packages/fxa-shared/payments/stripe.ts:
```diff
@@ -0,0 +1,493 @@
+  PLAY_SKU_IDS = 'playSkuIds',
+}
+
+export type StripeHelperConfig = {
+  subscriptions: {
+    stripeApiKey: string;
+    stripeWebhookSecret: string;
+    productConfigsFirestore: {
+      enabled: boolean;
+    };
+  };
+  authFirestore: {
+    prefix: string;
+  };
+  subhub: {
+    plansCacheTtlSeconds: number;
+  };
+  redis: any; // TODO
+};
```
Ah cool, this is what I had thought would be around for the other refactored Base classes. 👍🏻

bbangert commented on packages/fxa-shared/payments/stripe.ts:
```diff
@@ -0,0 +1,493 @@
+ */
+export abstract class StripeHelper {
+  /** Stripe instance used to communicate with stripe api. */
+  public abstract readonly stripe: Stripe;
```
This is being used publicly?

bbangert commented on packages/fxa-shared/sentry/index.ts:
```diff
@@ -1,3 +1,68 @@
+import { Hub, Severity } from '@sentry/types';
+
+/** Minimal type containing validation error fields that are used by routines. */
+export type ValidationError = {
```
Should these extend Error? I noticed some of these `*Error` objects are types (here), some are classes, and sometimes they extend `Error`. It's not clear when its what, or why.

dschom replied:
Hmm, these objects are not being 'thrown' so not sure if it matters. They were just adding a bit of type safety, so a message could be built up. That being said, the underlying objects being sent in are joi validations errors. I didn't realize fxa-shared had an existing joi dependency already, so I can just use that instead.

dschom replied:
> 

Thanks for the review. I know this is a rather large change set. I am working on addressing the feedback. Here's the main things I will address in a second push:

1. Hoist the entirety of UserManager.queryCurrentSubscriptions to fxa-shared, which will ensnare other things.
2. Fix a couple fields that got accidentally marked as public.
3. See if I can work around passing the sentry instance to reportValidationError and address ValidationError type there.

dschom replied:
Alright, this has been moved to fxa-shared/sentry/report-validation-error.ts, and the now references the joi type for ValidationError.

bbangert reviewed (changes requested):
Okey, got through it, sorry about the delay. Bunch of questions, mainly consistency related.

dschom commented:
> 

Thanks for the review. I'll be working on addressing the feedback. It primary falls into three groups:

1. Hoisting entirety of UserManager.queryCurrentSubscriptions to base class.
2. Deal with a few spots where something got marked public accidentally.
3. Address feedback in fxa-shared/sentry/index.ts. I'll see if I can avoid passing the instance.

dschom commented on packages/fxa-shared/sentry/report-validation-error.ts:
```diff
@@ -0,0 +1,56 @@
+import * as Sentry from '@sentry/node';
```
By breaking this out into its own file separate from index.ts, the context issue was resolved. We must be careful not to accidentally include this file in ./index.ts. Doing so could result in @sentry/node being required on the client side which completely breaks things. Additionally, once the file was separated the strange issue I had with mocks on sentry also cleared up.

bbangert reviewed (approved):
LGTM!
